ext/nnstreamer/extra/nnstreamer_grpc_flatbuf.cc

```diff
@@ -0,0 +1,252 @@
+{
+  GstDataQueueItem *item;
+
+  /* until flushing */
+  while (gst_data_queue_pop (queue_, &item)) {
+    Message<Tensors> tensors;
+
+    _get_tensors_from_buffer (GST_BUFFER (item->object), tensors);
+    writer->Write (tensors);
+
+    g_free (item);
+  }
+
+  return Status::OK;
+}
+
+/** @brief gRPC client thread */
+void
+ServiceImplFlatbuf::_client_thread ()
+{
+  ClientContext context;
+
+  if (direction_ == GRPC_DIRECTION_TENSORS_TO_BUFFER) {
+    Message<Empty> empty;
+
+    /* initiate the RPC call */
+    std::unique_ptr< ClientWriter<Message<Tensors>> > writer(
+        client_stub_->SendTensors (&context, &empty));
+
+    _write_tensors (writer.get ());
+
+    writer->WritesDone ();
+    /**
+     * TODO: The below incurs assertion failure but it seems like a bug.
+     * Let's check it later with the latest gRPC version.
+     *
+     * writer->Finish ();
+     */
+  } else if (direction_ == GRPC_DIRECTION_BUFFER_TO_TENSORS) {
+    MessageBuilder builder;
+
+    auto empty_offset = nnstreamer::flatbuf::CreateEmpty (builder);
+    builder.Finish (empty_offset);
+
+    /* initiate the RPC call */
+    std::unique_ptr< ClientReader<Message<Tensors>> > reader(
+        client_stub_->RecvTensors (&context, builder.ReleaseMessage <Empty> ()));
+
+    _read_tensors (reader.get ());
+
+    reader->Finish ();
+  } else {
+    g_assert (0); /* internal logic error */
+  }
+}
+
+/** @brief start gRPC server handling flatbuf */
+gboolean
+ServiceImplFlatbuf::start_server (std::string address)
+{
+  /* listen on the given address without any authentication mechanism */
+  ServerBuilder builder;
+  builder.AddListeningPort (address, grpc::InsecureServerCredentials(), &port_);
+  builder.RegisterService (this);
+
+  /* start the server */
+  server_worker_ = builder.BuildAndStart ();
+  if (server_worker_.get () == nullptr)
+    return FALSE;
+
+  return TRUE;
+}
+
+/** @brief start gRPC client handling flatbuf */
+gboolean
+ServiceImplFlatbuf::start_client (std::string address)
+{
+  /* create a gRPC channel */
+  std::shared_ptr<Channel> channel = grpc::CreateChannel(
+      address, grpc::InsecureChannelCredentials());
+
+  /* connect the server */
+  client_stub_ = TensorService::NewStub (channel);
+  if (client_stub_.get () == nullptr)
+    return FALSE;
+
+  client_worker_ = std::thread ([this] { this->_client_thread (); });
+
+  return TRUE;
+}
+
+/** @brief convert tensors to buffer */
+void
+ServiceImplFlatbuf::_get_buffer_from_tensors (Message<Tensors> &msg,
+    GstBuffer **buffer)
+{
+  const Tensors *tensors = msg.GetRoot ();
+  guint num_tensor = tensors->num_tensor ();
+  GstMemory *memory;
+
+  *buffer = gst_buffer_new ();
+
+  for (guint i = 0; i < num_tensor; i++) {
+    const Tensor * tensor = tensors->tensor ()->Get (i);
+    const void * data = tensor->data ()->data ();
+    gsize size = VectorLength (tensor->data ());
+
+    memory = gst_memory_new_wrapped (GST_MEMORY_FLAG_READONLY,
```

Comment:
also this requires g_free to release newly allocated mem block

Reply:
Updated